File: java/LinkedList/LinkedListCycle.java
package java.LinkedList;
import java.util.HashMap;

public class LinkedListCycle {
    /**
     * 141. Linked List Cycle
     * Easy
     * Topics: Linked List, Two Pointers, hashTable
     */
    public boolean hasCycle(ListNode head) {
        HashMap<ListNode, Integer> map = new HashMap<>();
        ListNode current = head;
        int index = 0;
        while (current != null) {
            if (map.containsKey(current)) {
                return true;
            } 
            map.put(current, index);
            current = current.next;
            index++;
            
        }
        return false;
    }
}

File: python/LinkedList/LinkedListCycle.py
"""
141. LinkedList Cycle
Difficulty: Easy
Topics: LinkedList, HashTable, Two Pointers
"""
import ListNode

def hasCycle(head: ListNode) -> bool:
    # Fast and slow pointer
    if not head:
        return False
    if not head.next:
        return False

    fast = head
    slow = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    
    return False